File: python/killer-sudoku-helper/killer_sudoku_helper.py
"""
Calculate the possible combinations of numbers of a given length that add up to a given total
that could be solutions for Killer Soduku positions
"""

# looks like I should have been using itertools.combinations.

def combinations(target, size, exclude):
    """
    Find all number combinations of a given length that add up to a target value
    excluding numbers that are already taken.
    :param target: The target amount to reach
    :param size: The number of digits that must used to add up to the target
    :param exclude: A list of numbers that can not be used (taken)
    :returns: A list containing lists of number (length=size) that add up to target empty list if no matches
    """
    final = []
    ret = ['dummy']
    disallowed = [*exclude]
    
    while len(ret) > 0:
        ret = combination_helper(target, size, [], disallowed)
        if len(ret) > 0:
            disallowed = [*disallowed, *ret]
            final.append(ret)
    return final

def combination_helper(target, size, current, exclude):
    """
    Find one combination of numbers of length size that add up to target, for 1-9 not 
    including number in the exclude list. 
    :param target: The number to add up to
    :param size: The number of digits that must be consumed
    :param current: The digits already under consideration
    :param exclude: Digits that cannot be used
    :returns: A list of the digits of length size that add up to target or an empty list if no match found.
    """
    current_sum = sum(current)
    current_len = len(current)
    
    if current_len > size:
        return []
    
    if current_sum > target:
        return []
    
    if current_sum == target:
        if current_len == size:
            return current
        return []
    
    possibilities = [counter + 1 for counter in range(9) if counter + 1 not in current and counter + 1 not in exclude]
    if len(possibilities) == 0:
        return []
    
    for possibility in possibilities:
        ret = combination_helper(target, size, [*current, possibility], exclude)
        if len(ret) == size and sum(ret) == target:
            return ret
    
    return []
    

File: go/killer-sudoku-helper/killer_sudoku_helper.go
// Killer Sudoku Helper exercise - Find combinations that satisfy requirements
package killersudokuhelper

import "slices"

// Calculate the combinations of values 1-9 that satisfy the requirements.
// @param sum: The accepted combinations should add up to this
// @param size: The number of items needed in the accepted combinations
// @param exclude: Slice of values that may not be in the accepted combinations.
// @returns: A slice of slices of integers for the accepted unique combinations
// where each combination is sorted smallest to largest.
func Combinations(sum, size int, exclude []int) [][]int {
	empty := make([]int, 0)
	return Helper(sum, size, empty, exclude)
}

// Does the actual combinations work recursively but with a working set.
// @param sum: The accepted combinations should add up to this
// @param size: The number of items needed in the accepted combinations
// @param working: Slice of int of what we have built so far.
// @param exclude: Slice of values that may not be in the accepted combinations.
// @returns: A slice of slices of integers for the accepted unique combinations
// where each combination is sorted smallest to largest.
func Helper(sum int, size int, working []int, exclude []int) [][]int {
	var result [][]int = make([][]int, 0)
	var currentSum int = 0
	for _, value := range working {
		currentSum += value
	}
	// Base case we are there
	if len(working) == size && currentSum == sum {
		slices.Sort(working)
		result = append(result, working)
		return result
	}
	// Another base case can't add any more.
	if len(working)+1 > size || currentSum > sum {
		return result
	}

	for i := 1; i <= 9; i++ {
		if currentSum+i > sum || slices.Contains(working, i) || slices.Contains(exclude, i) {
			// Too Big, Already used in partial solution, or was excluded.
			continue
		}
		var next []int = make([]int, len(working))
		copy(next, working)
		next = append(next, i)
		output := Helper(sum, size, next, exclude)
		if len(output) > 0 {
			for _, row := range output {
				slices.Sort(row)
				if !slices.ContainsFunc(result, func(slice []int) bool {
					if len(row) == len(slice) {
						for index, value := range slice {
							if row[index] != value {
								return false
							}
						}
						return true
					}
					return false
				}) {
					result = append(result, row)
				}
			}
		}
	}
	return result
}
